rust/src/witness_builder.rs

```diff
@@ -0,0 +1,415 @@
+        self.redeemers.insert(redeemer.clone());
+    }
+
+    pub fn add_all(&mut self, requirements: &RequiredWitnessSet) {
+        self.vkeys.extend(requirements.vkeys.iter().cloned());
+        self.bootstraps.extend(requirements.bootstraps.iter().cloned());
+        self.native_scripts.extend(requirements.native_scripts.iter().cloned());
+        self.plutus_scripts.extend(requirements.plutus_scripts.iter().cloned());
+        self.plutus_data.extend(requirements.plutus_data.iter().cloned());
+        self.redeemers.extend(requirements.redeemers.iter().cloned());
+    }
+
+    pub (crate) fn to_str(&self) -> String {
+        let vkeys = self.vkeys.iter().map(|key| format!("Vkey:{}", hex::encode(key.to_bytes()))).collect::<Vec<String>>().join(",");
+        let bootstraps = self.bootstraps.iter().map(|key| format!("Legacy Bootstraps:{}", hex::encode(key.to_bytes()))).collect::<Vec<String>>().join(",");
+        let native_scripts = self.native_scripts.iter().map(|hash| format!("Native script hash:{}", hex::encode(hash.to_bytes()))).collect::<Vec<String>>().join(",");
+        let plutus_scripts = self.plutus_scripts.iter().map(|hash| format!("Plutus script hash:{}", hex::encode(hash.to_bytes()))).collect::<Vec<String>>().join(",");
+        let plutus_data = self.plutus_data.iter().map(|hash| format!("Plutus data hash:{}", hex::encode(hash.to_bytes()))).collect::<Vec<String>>().join(",");
+        let redeemers = self.redeemers.iter().map(|key| format!("Redeemer:{}-{}", hex::encode(key.tag().to_bytes()), key.index().to_str())).collect::<Vec<String>>().join(",");
+
+        [vkeys, bootstraps, native_scripts, plutus_scripts, plutus_data, redeemers].iter().filter(|msg| msg.len() > 0).cloned().collect::<Vec<String>>().join("\n")
+    }
+
+    pub (crate) fn len(&self) -> usize {
+        self.vkeys.len() +
+            self.bootstraps.len() +
+            self.native_scripts.len() +
+            self.plutus_scripts.len() +
+            self.plutus_data.len() +
+            self.redeemers.len()
+    }
+
+    pub fn new() -> Self {
+        Self {
+            vkeys: HashSet::new(),
+            bootstraps: HashSet::new(),
+            native_scripts: HashSet::new(),
+            plutus_scripts: HashSet::new(),
+            plutus_data: HashSet::new(),
+            redeemers: HashSet::new(),
+        }
+    }
+}
+
+/// Builder de-duplicates witnesses as they are added 
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct TransactionWitnessSetBuilder {
+    // See Alonzo spec section 3.1 which defines the keys for these types
+    vkeys: HashMap<Vkey, Vkeywitness>,
+    bootstraps: HashMap<Vkey, BootstrapWitness>,
+    native_scripts: HashMap<ScriptHash, NativeScript>,
+    plutus_scripts: HashMap<ScriptHash, PlutusScript>,
+    plutus_data: HashMap<DataHash, PlutusData>,
+    redeemers: HashMap<RedeemerWitnessKey, Redeemer>,
+
+    // witnesses that need to be added for the build function to succeed
+    required_wits: RequiredWitnessSet,
```

Comment:
@SebastienGllmt , can you plz leave a lil comment, what's the idea with this field, i.e. how and when it is presumed to be used?